npc.cpp

```diff
@@ -108,31 +89,32 @@ void Game::GrowWMons(int rate)
 								 (ItemDefs[mon].flags & ItemType::DISABLED)))
 								avail = 1;
 
-							if (!avail) {
-								printf("\n\n NO AVAL  !!!!!!!!!!! \n\n");
-							}
-							if (avail)
+							if (avail) {
 								wanted += TerrainDefs[reg->type].wmonfreq;
+							}
 						}
 					}
 				}
 
-				printf("\n\n GrowWMons 2: %d \n\n", wanted);
-
 				wanted /= 10;
 				wanted -= mons;
 				wanted = (wanted*rate + getrandom(100))/100;
 
-				printf("\n\n GrowWMons 3: %d . mons: %d. wanted: %d. rate: %d \n\n", wanted, mons, wanted, rate);
-
 				if (wanted > 0) {
-					for (int i=0; i< wanted;) {
-						int m=getrandom(8);
-						int n=getrandom(8)*2+m%2;
+					// TODO: instead of loop guard need to check how many available regions
+					// are there and random them
+					int loop_guard = 1000;
+					for (int i=0; i < wanted;) {
+						int m = getrandom(8);
+						int n = getrandom(16)*2+m%2;
 						ARegion *reg = pArr->GetRegion(m+xsec*8, n+ysec*16);
-						if (reg && !reg->IsGuarded() && MakeWMon(reg)) {
+						if (reg && reg->zloc == level && !reg->IsGuarded() && MakeWMon(reg)) {
 							i++;
 						}
+
+						// If all level hexes guarded we need to prevent infinitie loop
```

Comment:
Actually, if all hexes in this section are guarded then wanted would be 0, so we don't need to worry about this.  Maybe my explanation of what can go wrong here was not clear?

Reply:
it was. Will rewrite